python/swig/libprio.i

```diff
@@ -0,0 +1,282 @@
+        // Free malloc'ed data from within PrioClient_encode
+        if (*$1) {
+            free(*$1);
+        }
+    }
+
+    %apply (unsigned char **, unsigned int *) {
+        (unsigned char **forServerA, unsigned int *aLen),
+        (unsigned char **forServerB, unsigned int *bLen)
+    }
+%enddef
+
+
+%define VERIFIER_SET_DATA()
+    // PrioVerifier_set_data
+    %typemap(in) (unsigned char * data, unsigned int dataLen) {
+        if (!PyBytes_Check($input)) {
+            PyErr_SetString(PyExc_ValueError, "Expecting a byte string");
+            SWIG_fail;
+        }
+        $1 = (unsigned char*) PyBytes_AsString($input);
+        $2 = (unsigned int) PyBytes_Size($input);
+    }
+%enddef
+
+
+%define TOTAL_SHARE_FINAL()
+    // PrioTotalShare_final
+    %typemap(in) (const_PrioConfig, unsigned long long *) {
+        $1 = PyCapsule_GetPointer($input, "PrioConfig");
+        $2 = malloc(sizeof(unsigned long long)*PrioConfig_numDataFields($1));
+    }
+
+    %typemap(argout) (const_PrioConfig, unsigned long long *) {
+        $result = SWIG_Python_AppendOutput(
+            $result,
+            PyByteArray_FromStringAndSize(
+                (const char*)$2,
+                sizeof(unsigned long long) * PrioConfig_numDataFields($1)
+            )
+        );
+        if ($2) {
+            free($2);
+        }
+    }
+    
+    %apply (const_PrioConfig, unsigned long long *) {
+        (const_PrioConfig cfg, unsigned long long *output)
+    }
+
+    // PrioTotalShare_final_uint
+    // Note the logic is similar to PrioClient_encode/
+    %typemap(in) (const_PrioConfig, unsigned long long **, unsigned int *) 
+        (unsigned long long *data, unsigned int len = 0)
+    {
+        $1 = PyCapsule_GetPointer($input, "PrioConfig");
+        $2 = &data;
+        $3 = &len;
+    }
+
+    %typemap(argout) (const_PrioConfig, unsigned long long **, unsigned int *) {
+        $result = SWIG_Python_AppendOutput(
+            $result,
+            PyByteArray_FromStringAndSize((const char*)*$2, *$3)
+        );
+        if (*$2) {
+            free(*$2);
+        }
+    }
+
+    %apply (const_PrioConfig, unsigned long long**, unsigned int*) {
+        (const_PrioConfig cfg, unsigned long long** outputWrap, unsigned int *outputWrapLen)
+    }
+
+    %ignore PrioTotalShare_final_uint;
+    %rename(PrioTotalShare_final_uint) PrioTotalShare_final_uint_wrapper;
+    %inline {
+        /// This wrapper exists because config is now a leaky abstraction.
+        /// Precision should be added directly into the configuration struct if
+        /// possible. Note that the arguments are reordered in order to keep the
+        /// same interface to PrioTotalShare_final. The precision cannot be used
+        /// within an output typemap. Instead, we dynamically allocate memory in
+        /// references to stack allocated swig variables (the typemap(in)).
+        SECStatus PrioTotalShare_final_uint_wrapper(
```

Comment:
I'm wondering if precision could have been added directly into the config object so it doesn't have to be passed out of band. I thought the opaque pointer idiom was meant to hide those details from the public interface?

Reply:
We made the decision to keep it out of band to maximize reuse of audited code and to keep full backwards compatibility.

Reply:
I see, it certainly makes sense if a subset of code paths are fully audited outside of the patch for b-bit integer support. Suppose the code had to be re-audited though. If the precision were added to the config struct, wouldn't the API still be fully backwards compatible?  The uint variants of functions would still need to exist, but the config would abstract the details away once instantiated since the library is only passing around references to the config.

Reply:
You would still have to go back to code that is using the old API and set precision, even though you don't want to use it in the bit vector case. Also this re-audit would need to happen.

Reply:
Ah right, the old API is equivalent to the uint one where the precision is set to 1. Thank you for clearing that up, I understand how this is intended to be used now.

Reply:
You're very welcome, and thanks for the bindings!